### PR TITLE
Build history enhancements

### PR DESCRIPTION
**Pull request by lefticus**
Implements KAGN-4-42506-05 task 6.11, enhanced build history on a per-branch basis.

### CODE REVIEW

File: ci.rb [hunk 1]
```diff
@@ -149,6 +155,63 @@
...
+    rescue SyntaxError => e
+      $logger.info("'#{e.message}' error while reading limits file")
+    rescue Psych::SyntaxError => e
+      $logger.info("'#{e.message}' error while reading limits file")
+    rescue => e
+      $logger.info("'#{e.message}' '#{e.backtrace}' error while reading limits file")
+    end
+
+    return {};
+  }.();
```
Comment (Myoldmopar) [inline ci.rb]: What's the advantage to using a lambda here and not just executing the code?

File: cleanup.rb [hunk 2]
```diff
@@ -46,12 +90,24 @@ def clean_up(client, repository, results_repository, results_path, age_limit)
...
+
+  # These are the absolute failsafe limits required by github
+  # if we are approaching 1000 files in a single directory
   if files.size > 800
     if files.size > 999
       branch_history_limit = 1
       file_age_limit = age_limit + 1
     else
-      branch_history_limit = 5
+      branch_history_limit = [5, branch_history_limit].min
```
Comment (Myoldmopar) [inline cleanup.rb]: So this overrides the limit to a maximum of 5 results, even for long running branches, if the files in the directory are approaching 1000?

Comment (lefticus) [inline cleanup.rb]: Yes, this is the failsafe that comes into play if any one folder is approaching 1000 files in that folder, we don't want the >5000 repository number (which may be smaller) to accidentally be overridden by a potentially larger failsafe number. So we take the min, which should be the safest option.

**Comment by lefticus** [comment]
@Myoldmopar  when would be a good day to merge this and deploy it - on the off chance it causes issues? Perhaps I should merge it before coming down to Golden next week to see you?

**Comment by Myoldmopar** [comment]
I'd say now is actually probably the best time. We just got done with a release milestone, so while there will be some bug fixes going in over the next few weeks, we have some time right now to deal with an oops moment if it pops up. And yes we will still chat on everything next Monday.

**Comment by lefticus** [comment]
Ok, I will do that around lunch time.

**Pull request merged**